java-frontend/src/main/java/org/sonar/java/resolve/LeastUpperBound.java

```diff
@@ -0,0 +1,336 @@
+
+    return leastContainingParameterization(reducedList);
+  }
+
+  private Type leastContainingTypeArgument(JavaType type1, JavaType type2) {
+    Preconditions.checkArgument(type1.isTagged(JavaType.PARAMETERIZED) && type2.isTagged(JavaType.PARAMETERIZED));
+
+    TypeSubstitution typeSubstitution1 = ((ParametrizedTypeJavaType) type1).typeSubstitution;
+    TypeSubstitution typeSubstitution2 = ((ParametrizedTypeJavaType) type2).typeSubstitution;
+
+    TypeSubstitution newTypeSubstitution = new TypeSubstitution();
+    for (TypeVariableJavaType typeVar : typeSubstitution1.typeVariables()) {
+      JavaType subs1 = typeSubstitution1.substitutedType(typeVar);
+      JavaType subs2 = typeSubstitution2.substitutedType(typeVar);
+
+      JavaType newSubs = getNewTypeArgumentType(subs1, subs2);
+      newTypeSubstitution.add(typeVar, newSubs);
+    }
+
+    return parametrizedTypeCache.getParametrizedTypeType(type1.symbol, newTypeSubstitution);
+  }
+
+  private JavaType getNewTypeArgumentType(JavaType subs1, JavaType subs2) {
+    boolean isWildcard1 = subs1.isTagged(JavaType.WILDCARD);
+    boolean isWildcard2 = subs2.isTagged(JavaType.WILDCARD);
+
+    JavaType result;
+    if (subs1.equals(subs2)) {
+      result = subs1;
+    } else if (isWildcard1 && isWildcard2) {
+      result = lctaBothWildcards((WildCardType) subs1, (WildCardType) subs2);
+    } else if (isWildcard1 ^ isWildcard2) {
+      JavaType rawType = isWildcard1 ? subs2 : subs1;
+      WildCardType wildcardType = (WildCardType) (isWildcard1 ? subs1 : subs2);
+      result = lctaOneWildcard(rawType, wildcardType);
+    } else {
+      result = lctaNoWildcard(subs1, subs2);
+    }
+    return result;
+  }
+
+  private JavaType lctaOneWildcard(JavaType rawType, WildCardType wildcardType) {
+    if (wildcardType.boundType == WildCardType.BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Sets.newHashSet(rawType, wildcardType.bound));
+      return parametrizedTypeCache.getWildcardType(glb, WildCardType.BoundType.SUPER);
+    }
+    JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(rawType, wildcardType.bound));
+    return parametrizedTypeCache.getWildcardType(lub, WildCardType.BoundType.EXTENDS);
+  }
+
+  private JavaType lctaBothWildcards(WildCardType type1, WildCardType type2) {
+    if (type1.boundType == WildCardType.BoundType.SUPER && type2.boundType == WildCardType.BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Sets.newHashSet(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(glb, WildCardType.BoundType.SUPER);
+    }
+    if (type1.boundType == WildCardType.BoundType.EXTENDS && type2.boundType == WildCardType.BoundType.EXTENDS) {
+      JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(lub, WildCardType.BoundType.EXTENDS);
+    }
+    if (type1.bound == type2.bound) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)